bindings/haskell/test/BasicTest.hs

```diff
@@ -27,25 +28,59 @@ basicTests :: TestTree
 basicTests =
   testGroup
     "Basic Tests"
-    [ testCase "testBasicOperation" testBasicOperation
+    [ testCase "testBasicOperation" testRawOperation
+    , testCase "testMonad" testMonad
+    , testCase "testError" testError
     ]
 
-testBasicOperation :: Assertion
-testBasicOperation = do
+testRawOperation :: Assertion
+testRawOperation = do
   Right op <- newOp "fs" $ HashMap.fromList [("root", "/tmp/opendal-test")]
-  writeOp op "key1" "value1" >>= (@?= Right ())
-  writeOp op "key2" "value2" >>= (@?= Right ())
-  readOp op "key1" >>= (@?= Right "value1")
-  readOp op "key2" >>= (@?= Right "value2")
-  isExistOp op "key1" >>= (@?= Right True)
-  isExistOp op "key2" >>= (@?= Right True)
-  createDirOp op "dir1/" >>= (@?= Right ())
-  isExistOp op "dir1/" >>= (@?= Right True)
-  copyOp op "key1" "key3" >>= (@?= Right ())
-  isExistOp op "key3" >>= (@?= Right True)
-  isExistOp op "key1" >>= (@?= Right True)
-  renameOp op "key2" "key4" >>= (@?= Right ())
-  isExistOp op "key4" >>= (@?= Right True)
-  isExistOp op "key2" >>= (@?= Right False)
-  deleteOp op "key1" >>= (@?= Right ())
-  isExistOp op "key1" >>= (@?= Right False)
\ No newline at end of file
+  writeOpRaw op "key1" "value1" >>= (@?= Right ())
+  writeOpRaw op "key2" "value2" >>= (@?= Right ())
+  readOpRaw op "key1" >>= (@?= Right "value1")
+  readOpRaw op "key2" >>= (@?= Right "value2")
+  isExistOpRaw op "key1" >>= (@?= Right True)
+  isExistOpRaw op "key2" >>= (@?= Right True)
+  createDirOpRaw op "dir1/" >>= (@?= Right ())
+  isExistOpRaw op "dir1/" >>= (@?= Right True)
+  copyOpRaw op "key1" "key3" >>= (@?= Right ())
+  isExistOpRaw op "key3" >>= (@?= Right True)
+  isExistOpRaw op "key1" >>= (@?= Right True)
+  renameOpRaw op "key2" "key4" >>= (@?= Right ())
+  isExistOpRaw op "key4" >>= (@?= Right True)
+  isExistOpRaw op "key2" >>= (@?= Right False)
+  deleteOpRaw op "key1" >>= (@?= Right ())
+  isExistOpRaw op "key1" >>= (@?= Right False)
+
+testMonad :: Assertion
+testMonad = do
+  Right op <- newOp "fs" $ HashMap.fromList [("root", "/tmp/opendal-test2")]
+  runOp op operation >>= (@?= Right ())
+ where
+  operation = do
+    writeOp "key1" "value1"
+    writeOp "key2" "value2"
+    readOp "key1" >>= liftIO . (@?= "value1")
+    readOp "key2" >>= liftIO . (@?= "value2")
+    isExistOp "key1" >>= liftIO . (@?= True)
+    isExistOp "key2" >>= liftIO . (@?= True)
+    createDirOp "dir1/"
+    isExistOp "dir1/" >>= liftIO . (@?= True)
+    copyOp "key1" "key3"
+    isExistOp "key3" >>= liftIO . (@?= True)
+    isExistOp "key1" >>= liftIO . (@?= True)
+    renameOp "key2" "key4"
+    isExistOp "key4" >>= liftIO . (@?= True)
+    isExistOp "key2" >>= liftIO . (@?= False)
+    deleteOp "key1"
+    isExistOp "key1" >>= liftIO . (@?= False)
+
+testError :: Assertion
+testError = do
+  Right op <- newOp "fs" $ HashMap.fromList [("root", "/tmp/opendal-test3")]
```

Comment:
I prefer to test again memory service to avoid affected by env. For example, this test will not pass on windows.

Reply:
updated in new commit